docs/rfcs/025-generation-numbers.md

```diff
@@ -0,0 +1,468 @@
+
+Eventually, node 0 will somehow realize it should stop running by some means:
+
+- A hard-stop of the VM it is running on
+- It tries to communicate with another peer that sends it an error response
+  indicating its generation is out of date
+- It tries to do some deletions, and discovers when synchronizing with the
+  control plane that its generation number is stale.
+
+#### Case B: direct node replacement with blank node (cold standby)
+
+1. Let's say node 0 fails, and there may be some other peers but they aren't relevant.
+1. Some external mechanism notices that the node is unavailable, and creates
+   a "new node 0" (Node 0b) which is a physically separate server. The original node 0
+   (Node 0a) may still be running.
+1. On startup, node 0b acquires a new generation number. It doesn't have any local state,
+   so the control plane must send attach requests to node 0b for all the tenants. However,
+   these do not have to increment the attachment generation as they're still attached to
+   logical node 0, it's just a different physical node. So we have an O(tenant_count) communication
+   to the pageserver, but not an O(tenant_count) write to the database.
+1. S3 writes continue from nodes 0a and 0b, but the writes do not collide due to different
+   node generation in the suffix, and the writes from node 0a are not visible to the rest
+   of the system because endpoints are reading only from node 0b.
+
+# Appendix B: interoperability with other features
+
+## Sharded Keyspace
+
+The design in this RFC maps neatly to a sharded keyspace design where subsets of the key space
+for a tenant are assigned to different pageservers:
+
+- the "unit of work" for attachments becomes something like a TenantShard rather than a Tenant
+- TenantShards get generation numbers just as Tenants do.
+- Write workload (ingest, compaction) for a tenant is spread out across pageservers via
+  TenantShards, but each TenantShard still has exactly one valid writer at a time.
+
+## Read replicas
+
+For historical reads to LSNs below the remote persistent LSN, any node may act as a reader at any
+time: remote data is logically immutable data, and the use of deferred deletion in this RFC helps
+mitigate the fact that remote data is not _physically_ immutable (i.e. the actual data for a given
+page moves around as compaction happens).
+
+A read replica needs to be aware of generations in remote data in order to read the latest
+metadata (find the index_part.json with the latest suffix). This does not have to involve any
+control plane interaction: a read replica node can scan remote index files to find the latest one. However,
+it may make sense to send notifications from the control plane to avoid the need for read replicas
+to poll S3 to detect when the writer for a particular tenant changed, or to expose this information
+via some intermediate pub/sub mechanism for async consumption by read replicas.
```

Comment:
> intermediate pub/sub mechanism

We have this thing called `broker`. It lacks persistence and is quite primitive, but works for simple pub/sub and may still be useful for some scenarios outside its current usage area

Reply:
This is an excellent point.  I have taken advantage of the storage broker a few places in the latest cut of this doc -- publishing generations in messages to the safekeeper (and then carrying them onward into the broker) is useful for read replicas, and also for new attachments to learn the previous generation to load its index.